software/controller/lib/hal/rx_buf_uart_dma.h

```diff
@@ -0,0 +1,65 @@
+/* Copyright 2020-2021, RespiraWorks
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#pragma once
+
+#include "network_protocol.pb.h"
+#include "uart_dma.h"
+
+// RX buffer wrapper that controls restart of reception and provides count of received bytes.
+// We need this class to abstract hardware access from FrameDetector FSM, so it can be used in GUI
+template <uint32_t RxBytesMax>
+class RxBufferUartDma {
+ public:
+  explicit RxBufferUartDma(UartDma &uart_dma) : uart_dma_(uart_dma) {}
+
+  // Sets up underlying receive infrastructure and starts the first reception
+  [[nodiscard]] bool begin(RxListener *rxl) {
+    uart_dma_.enable_character_match();
+    return uart_dma_.start_rx(rx_buffer_, RxBytesMax, rxl);
+  }
+
+  // Restarts the ongoing reception, this means the rx_buffer_ will be written from the beginning
+  void restart_rx(RxListener *rxl) {
+    uart_dma_.stop_rx();
+    [[maybe_unused]] bool started = uart_dma_.start_rx(rx_buffer_, RxBytesMax, rxl);
+  }
+
+  // Returns how many bytes were written into rx_buf
+  uint32_t received_length() { return (RxBytesMax - uart_dma_.rx_bytes_left()); }
+
+  // Returns the rx_buffer_
+  uint8_t *get() { return rx_buffer_; }
+
+#ifdef TEST_MODE
+  // Puts a byte to rx_buffer_
+  void test_put_byte(uint8_t b);
+#endif
+
+ private:
+  // UartDma provides receiving to this buffer
+  UartDma &uart_dma_;
+  // Buffer into which the data is received
+  uint8_t rx_buffer_[RxBytesMax] = {0};
+};
+
+#ifdef TEST_MODE
+extern uint32_t rx_index;
```

Comment:
I kind of don't like this testing stuff to live with the classes, especially as it is all ifdeffed, which just messes with coverage statistics and overall makes things hard to follow.

Maybe we could make the private members protected and make the test code derive a hacked version of it for mocking. I am sure there are even more "correct" ways of doing such mocking, but this would be pretty cheap and good enough for now?

Reply:
This variable is only finally used in unit tests for the reimplemented `Comms` code in the last PR of this chain. This is a type of invasive mocking. I have made it a member (yet public) variable inside an `ifdef` block. Once we get to that final PR, we should dig down into this class explicitly not by use of globals.

Unfortunately, we cannot do the same as with `FrameDetector` and implement this in a derived mock class because of how we use UartDma. A number of lower-level modules must be made mockable before we can do something about this.